contracts/modules/v2/StakingPool.sol

```diff
@@ -173,145 +174,238 @@ contract StakingPool is IStakingPool, ERC721 {
-    }
+    uint newStakeShares = _stakeSharesSupply == 0
+      ? sqrt(amount)
+      : _stakeSharesSupply * amount / _activeStake;
 
-    /* store */
+    uint newRewardsShares = calculateRewardSharesAmount(newStakeShares, groupId);
 
     group.stakeShares += newStakeShares;
     group.rewardsShares += newRewardsShares;
-    group.groupSharesSupply += newGroupShares;
 
-    stakeGroups[groupId] = group;
-    balanceOf[positionId][groupId] += newGroupShares;
+    /* storage writes */
 
+    // update globals
     activeStake = _activeStake + amount;
     stakeSharesSupply = _stakeSharesSupply + newStakeShares;
     rewardsSharesSupply = _rewardsSharesSupply + newRewardsShares;
+
+    // update group and staker data
+    groups[groupId] = group;
+    positions[tokenId][groupId] = position;
+    pendingRewards[tokenId] = pendingReward;
+
+    /* end storage writes */
   }
 
-  function withdraw(WithdrawParams[] memory params) external {
-    // 1. check nft ownership / allowance
-    // 2. loop through each group:
-    // 2.1. check group expiration if the deposit flag is set
-    // 2.2. calculate the reward amount if the reward flag is set
-    // 3. sum up nxm amount of each group
-    // 4. transfer nxm to staker
+  function calculateRewardSharesAmount(
+    uint stakeSharesAmount,
+    uint groupId
+  ) internal view returns (uint) {
+
+    uint lockDuration = (groupId + 1) * GROUP_SIZE - block.timestamp;
+    uint maxLockDuration = GROUP_SIZE * 8;
+
+    // TODO: determine extra rewards formula
+    return
+      stakeSharesAmount
+      * REWARDS_SHARES_RATIO
+      * lockDuration
+      / REWARDS_SHARES_DENOMINATOR
+      / maxLockDuration;
+  }
+
+  function withdraw(WithdrawParams[] calldata params) external {
+
+    updateGroups();
+
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+    uint _firstActiveGroupId = block.timestamp / GROUP_SIZE;
+
+    for (uint i = 0; i < params.length; i++) {
+
+      uint stakeToWithdraw;
+      uint rewardToWithdraw;
+
+      uint tokenId = params[i].tokenId;
+      uint groupCount = params[i].groupIds.length;
+
+      for (uint j = 0; j < groupCount; j++) {
+
+        uint groupId = params[i].groupIds[j];
+        Position memory position = positions[tokenId][groupId];
+
+        // can withdraw stake only if the group is expired
+        if (params[i].withdrawStake && groupId < _firstActiveGroupId) {
+
+          // calculate the amount of nxm for this position
+          uint stake = expiredGroups[groupId].stakeAmountAtExpiry;
+          uint stakeShareSupply = expiredGroups[groupId].stakeShareSupplyAtExpiry;
+          stakeToWithdraw += stake * position.stakeShares / stakeShareSupply;
+
+          // mark as withdrawn
+          position.stakeShares = 0;
+        }
+
+        if (params[i].withdrawRewards) {
+
+          // calculate reward since checkpoint
+          uint newRewardPerShare = _accNxmPerRewardsShare - position.lastAccNxmPerRewardShare;
```

Comment:
Cannot use "per share" because we have to specify the share type and in this case it's rewards share, which would result in "accumulated reward per reward share" which makes it more confusing.

Reply:
No. Explained below.